src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs

```diff
@@ -100,24 +99,95 @@ public abstract class JsonPropertyInfo
+                _shouldSerializeIsExplicitlySet = false;
+            }
+        }
+
+        private Func<object, object?, bool>? _shouldSerialize;
         private bool _shouldSerializeIsExplicitlySet;
+        private JsonIgnoreCondition? _ignoreCondition;
+
+        private protected abstract Func<object, object?, bool> GetShouldSerializeForIgnoreCondition(JsonIgnoreCondition condition);
+
+        /// <summary>
+        /// Gets or sets a custom attribute provider for the current property.
+        /// </summary>
+        /// <remarks>
+        /// When resolving metadata via <see cref="DefaultJsonTypeInfoResolver"/> this
+        /// will be populated with the underlying <see cref="MemberInfo" /> of the serialized property or field.
+        ///
+        /// Setting a custom attribute provider will have no impact on the contract model,
+        /// but serves as metadata for downstream contract modifiers.
+        /// </remarks>
+        internal ICustomAttributeProvider? AttributeProvider
+        {
+            get => _attributeProvider;
+            set
+            {
+                VerifyMutable();
+
+                _attributeProvider = value;
+            }
+        }
+
+        private ICustomAttributeProvider? _attributeProvider;
+        internal string? MemberName { get; private protected set; }
+        internal MemberTypes MemberType { get; private protected set; }
+        internal bool IsVirtual { get; private set; }
+
+        /// <summary>
+        /// Specifies whether the current property is a special extension data property.
+        /// </summary>
+        /// <remarks>
+        /// Properties annotated with <see cref="JsonExtensionDataAttribute"/>
+        /// will appear here when using <see cref="DefaultJsonTypeInfoResolver"/> or <see cref="JsonSerializerContext"/>.
+        /// </remarks>
+        internal bool IsExtensionData
+        {
+            get => _isExtensionDataProperty;
+            set
+            {
+                VerifyMutable();
+
+                if (value && !JsonTypeInfo.IsValidExtensionDataProperty(this))
+                {
+                    ThrowHelper.ThrowInvalidOperationException_SerializationDataExtensionPropertyInvalid(this);
+                }
+
+                _isExtensionDataProperty = value;
+            }
+        }
+
+        private bool _isExtensionDataProperty;
 
-        internal JsonPropertyInfo(JsonTypeInfo? parentTypeInfo)
+        internal JsonPropertyInfo(Type declaringType, Type propertyType, JsonTypeInfo? parentTypeInfo, JsonSerializerOptions options)
         {
-            // null parentTypeInfo means it's not tied yet
-            ParentTypeInfo = parentTypeInfo;
+            DeclaringType = declaringType;
+            PropertyType = propertyType;
+            Options = options;
+            ParentTypeInfo = parentTypeInfo; // null parentTypeInfo means it's not tied yet
```

Comment:
should we debug assert that if parentTypeInfo is not null then `declaringType == parentTypeInfo.Type`?

Reply:
We could, I'll add in the next round of PRs